r2i/tflite/engine.cc

```diff
@@ -193,41 +198,120 @@ std::shared_ptr<r2i::IPrediction> Engine::Predict (std::shared_ptr<r2i::IFrame>
     return nullptr;
   }
 
-  // Check the model quantization, only 32 bits allowed
-  if (kTfLiteFloat32 != interpreter->tensor(input)->type) {
-    error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
-               "The provided model quantization is not allowed, only float32 is supported");
-    return nullptr;
-  }
-
-  auto input_tensor = this->interpreter->typed_tensor<float>(input);
-  auto input_data = (float *)frame->GetData();
-
-  if (!input_data) {
-    error.Set (RuntimeError::Code::FRAMEWORK_ERROR, "Failed to get image data");
+  this->PreprocessInputData(static_cast<float *>(frame->GetData()),
+                            wanted_width * wanted_height * wanted_channels, this->interpreter.get(), error);
+  if (r2i::RuntimeError::EOK != error.GetCode()) {
     return nullptr;
   }
 
-  memcpy(input_tensor, input_data,
-         wanted_height * wanted_width * wanted_channels * sizeof(float));
-
   if (this->interpreter->Invoke() != kTfLiteOk) {
     error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
                "Failed to invoke tflite!");
     return nullptr;
   }
 
-  int output = this->interpreter->outputs()[0];
-  TfLiteIntArray *output_dims = this->interpreter->tensor(output)->dims;
-  auto output_size = GetRequiredBufferSize(output_dims) * sizeof(float);
-  auto *tensor_data = this->interpreter->typed_output_tensor<float>(0);
+  int output_size;
+  auto tensor_data = this->GetOutputTensorData(this->interpreter.get(),
+                     output_size, error);
+  if (r2i::RuntimeError::EOK != error.GetCode()) {
+    return nullptr;
+  }
+
   prediction->SetTensorValues(tensor_data, output_size);
 
   return prediction;
 }
 
 Engine::~Engine () {
   this->Stop();
+  this->interpreter.reset();
+}
+
+void Engine::SetupResolver(::tflite::ops::builtin::BuiltinOpResolver
+                           &/*resolver*/) {
+  // No implementation for tflite engine
+}
+
+void Engine::SetInterpreterContext(::tflite::Interpreter */*interpreter*/) {
+  // No implementation for tflite engine
+}
+
+void Engine::PreprocessInputData(const float *input_data, const int size,
+                                 ::tflite::Interpreter *interpreter, r2i::RuntimeError &error) {
+  const auto &input_indices = interpreter->inputs();
+  const auto *tensor = interpreter->tensor(input_indices[0]);
+
+  if (!input_data) {
+    error.Set (RuntimeError::Code::WRONG_API_USAGE, "Failed to get image data");
+    return;
+  }
+
+  if (kTfLiteUInt8 == tensor->type) {
+    auto *input_fixed_tensor = interpreter->typed_input_tensor<uint8_t>(0);
+
+    std::vector<uint8_t> input_data_fixed;
+    input_data_fixed.resize(size);
+    for (int index = 0; index < size; index++) {
+      input_data_fixed[index] = (uint8_t)input_data[index];
```

Comment:
This is just truncating the input from float to uint8. We need to apply the `zero` and `scale` here as well.